### DISCUSSION TITLE
Generic compose function to create a new model class with additional props from a base model

### DISCUSSION THREAD

**Discussion by evelant** [xaviergonz/mobx-keystone#295] [Q&A]
I'm trying to write a function to create many models that all extend from the same base model but have a couple additional properties. I have a large union type to model so I want to make it easy to create each individual case. I'm struggling to figure out how to do this in a generic way. I don't think I understand some deeper parts of typescript well enough to figure out why this isn't working.

Here's a shot at a generic function to make a new model from a base model but with potentially added props:
```
function compose<T extends string, U extends AnyModel, V extends ModelProps>(name: T, base: U, props: V) {
    @model(name)
    class ComposedClass extends ExtendedModel(() => ({ baseModel: base, props })) {}

    return ComposedClass
}
```

This results in the following error:

> Base constructor return type 'AnyModel & BaseModel<ModelPropsToData<V>, ModelPropsToCreationData<V>, any> & Omit<ModelPropsToData<V>, BaseModelKeys> & ModelPropsToSetter<...>' is not an object type or intersection of object types with statically known members.ts(2509)

It was intended to be used like so:

```
const myBigUnionType = tProp(
    types.array(
        types.or(
            compose("item1", MyBaseModel, {type: tProp(types.literal("option1"), "option1")}),
            compose("item2", MyBaseModel, {type: tProp(types.literal("option2"), "option2"), otherStuff: tProp(types.number, 0)}),
            ......
            compose("item26", MyBaseModel, {type: tProp(types.literal("option26"), "option26"), somethingElse: tProp(types.string)}),
)))
```

I can't find any way to add generic props to an ExtendedModel. It seems the only way to model this right now is to type out a separate class for each case of the union or forego type checking and optional values.

**Comment by xaviergonz** [comment]
Does this section of the docs help? https://mobx-keystone.js.org/classModels#factory-pattern--generics

**Reply by evelant** [reply]
That's similar to what I was trying to accomplish but not quite. In this case I wanted a function that returned a new class extended from `ExtendedModel` but where I could also pass extra prop definitions into the function to be added to the new class. It's passing in the prop definitions that TS has an issue with and I'm not quite sure why.

For example this works, adding a specific static prop in a function to a base model:

```
function createClassWithPropX<T extends string, U extends AbstractModelClass<AnyModel> >(name: T, base: U) {
    @model(name)
    class MyNewClass extends ExtendedModel(() => ({ baseModel: base, props: {x: prop<number>()} })) {}

    return MyNewClass
}
```

but this does not 

```
function compose<T extends string, U extends AbstractModelClass<AnyModel>, V extends ModelProps>(name: T, base: U, props: V) {
    @model(name)
    //
    class ComposedClass extends ExtendedModel(() => ({ baseModel: base, props })) {}

    return ComposedClass
}
```

it results in 

> Base constructor return type 'AnyModel & BaseModel<ModelPropsToData<V>, ModelPropsToCreationData<V>, any> & Omit<ModelPropsToData<V>, BaseModelKeys> & ModelPropsToSetter<...>' is not an object type or intersection of object types with statically known members.ts(2509)

I was originally trying to do this because I needed to create a bunch of similar models for a union type but each of them might have 1 or 2 additional props. This can be worked around by simply typing out separate classes for each union case but IMO this sort of general case compose function is an interesting problem and might be useful if it could work.
